File: jmlparser-jml-tools/src/test/java/com/github/jmlparser/wd/WDVisitorExprTest.java
package com.github.jmlparser.wd;

import com.github.javaparser.JavaParser;
import com.github.jmlparser.SolverTest;
import org.junit.jupiter.api.Assumptions;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvFileSource;

import static com.github.jmlparser.wd.WellDefinednessMain.isWelldefined;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * @author Alexander Weigl
 * @version 1 (14.06.22)
 */
class WDVisitorExprTest {
    private final JavaParser parser = new JavaParser();

    @ParameterizedTest
    @CsvFileSource(resources = "wd-expr.csv", delimiterString = "§")
    void wdExpression(String expr) {
        Assumptions.assumeTrue(SolverTest.z3Installed());
        assertTrue(isWelldefined(parser, expr));
    }

    @ParameterizedTest
    @CsvFileSource(resources = "not-wd-expr.csv", delimiterString = "§")
    void wdExpressionError(String expr) {
        Assumptions.assumeTrue(SolverTest.z3Installed());
        assertFalse(isWelldefined(parser, expr));
    }
}

File: tools/wd/src/test/kotlin/WDVisitorExprTest.kt
import com.github.javaparser.JavaParser
import io.github.jmltoolkit.smt.Z3
import io.github.jmltoolkit.wd.WdFacade.isWelldefined
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Assumptions
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.CsvFileSource

/**
 * @author Alexander Weigl
 * @version 1 (14.06.22)
 */
internal class WDVisitorExprTest {
    private val parser = JavaParser()

    @ParameterizedTest
    @CsvFileSource(resources = ["wd-expr.csv"], delimiterString = "§")
    fun wdExpression(expr: String) {
        Assumptions.assumeTrue(Z3.z3Installed())
        Assertions.assertTrue(isWelldefined(parser, expr))
    }

    @ParameterizedTest
    @CsvFileSource(resources = ["not-wd-expr.csv"], delimiterString = "§")
    fun wdExpressionError(expr: String) {
        Assumptions.assumeTrue(Z3.z3Installed())
        Assertions.assertFalse(isWelldefined(parser, expr))
    }
}
